### PR TITLE
[mlir] Add apply_patterns.linalg.pad_vectorization TD Op

### PR DESCRIPTION
**Pull request by banach-space**
This PR simply wraps `populatePadOpVectorizationPatterns` into a new
Transform Dialect Op: `apply_patterns.linalg.pad_vectorization`.

This change makes it possible to run (and test) the corresponding
patterns without:

  `transform.structured.vectorize_children_and_apply_patterns`.

Note that the Op above only supports non-masked vectorisation (i.e. when
the inputs are static), so, effectively, only fixed-width vectorisation
(as opposed to scalable vectorisation). As such, this change is required
to construct vectorization pipelines for tensor.pad targeting scalable
vectors.

To test the new Op and the corresponding patterns, I added
"vectorization-pad-patterns.mlir" - most tests have been extracted from
"vectorization-with-patterns.mlir".

As a side note, I feel that we should move
`GenericPadOpVectorizationPattern` out of
`populatePadOpVectorizationPatterns` as that's a "lower tensor.pad"
rather than a "vectorize tensor.pad" pattern. I am leaving that as a
TODO.

### CODE REVIEW

File: mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td [hunk 1]
```diff
@@ -84,6 +84,23 @@ def ApplyFoldAddIntoDestPatternsOp : Op<Transform_Dialect,
   let assemblyFormat = "attr-dict";
 }
 
+def ApplyPadVectorizationPatternsOp : Op<Transform_Dialect,
+    "apply_patterns.linalg.pad_vectorization",
+    [DeclareOpInterfaceMethods<PatternDescriptorOpInterface>]> {
+  let description = [{
+    Apply patterns that take tensor.pad and rewrites it as
+    vector.transfer_read/vector.transfer_write Ops.
+
```
Comment (javedabsar1) [inline mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:93]: The sentences seem broken as two independent thoughts are packed together. My attempt to refine (maybe lost in translation) - 
//  These patterns will either fold `tensor.pad` with an existing  `vector.transfer_read` or `vector.transfer_write`.  Or rewrite // it together with `tensor.insert_slice`  as a `vector.transfer_read +  vector.transfer_write` pair.
 Then explain about producer/consumer point

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp [hunk 2]
```diff
@@ -2285,7 +2285,7 @@ static SmallVector<Value> ofrToIndexValues(RewriterBase &rewriter, Location loc,
   return result;
 }
 
-/// Rewrite a tensor::PadOp into a sequence of EmptyOp, FillOp and
+/// Rewrite a tensor::PadOp into a sequence of EmptyOp, FillOp/GenerateOp and
```
Comment (javedabsar1) [inline mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:2288]: as a sequence of 
[ FillOp GenerateOp InsertSlice ]  .

File: mlir/test/Dialect/Linalg/vectorization-pad-patterns.mlir [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+// CHECK-LABEL: func @pad_and_transfer_write_dynamic_static
+//  CHECK-SAME:     %[[ARG0:.*]]: tensor<?x?xf32>, %[[SIZE:.*]]: index, %[[PADDING:.*]]: index
+//   CHECK-NOT:   tensor.pad
+//       CHECK:   %[[C0:.*]] = arith.constant 0 : index
+//       CHECK:   %[[SUB:.*]] = tensor.extract_slice %[[ARG0]][0, 0] [%[[SIZE]], 6] [1, 1] : tensor<?x?xf32> to tensor<?x6xf32>
+//       CHECK:   %[[VEC0:.*]] = call @make_vector() : () -> vector<7x9xf32>
+//       CHECK:   %[[RESULT:.*]] = vector.transfer_write %[[VEC0]], %[[SUB]][%[[C0]], %[[C0]]] : vector<7x9xf32>, tensor<?x6xf32>
+//       CHECK:   return %[[RESULT]]
+func.func @pad_and_transfer_write_dynamic_static(
```
Comment (javedabsar1) [inline mlir/test/Dialect/Linalg/vectorization-pad-patterns.mlir:89]: what is 'dynamic_static'  I guess you mean one part is dynamic and another is static ?

**Review by javedabsar1 — CHANGES_REQUESTED** [review]

File: mlir/test/Dialect/Linalg/vectorization-pad-patterns.mlir (hunk 3 above)
Comment (banach-space) [reply mlir/test/Dialect/Linalg/vectorization-pad-patterns.mlir:89]: I dunno :) This is just a copy of the original test and it wasn't me who wrote it.

Looking at the example, I assume this is referring to the "low" and "high" fields in the `tensor.pad` Op, which is the key thing in this test. Let me update these names accordingly.

File: mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td (hunk 1 above)
Comment (banach-space) [reply mlir/include/mlir/Dialect/Linalg/TransformOps/LinalgTransformOps.td:93]: Thank you for your suggestion - let me adapt it a bit. Hopefully with our joint efforts things will be clearer :) 

I will work under the assumption that we want to remove `GenericPadOpVectorizationPattern>` from this list - it doesn't really belong here - it's not a "vectorization" pattern per se. I am already in the process of refactoring it:
* https://github.com/llvm/llvm-project/pull/111349

One step at a time though!

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp (hunk 2 above)
Comment (banach-space) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:2288]: Let me revert this change altogether, I will be refactoring this in #111349 and later.

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp [hunk 4]
```diff
@@ -2712,6 +2712,9 @@ struct PadOpVectorizationWithInsertSlicePattern
 
 void mlir::linalg::populatePadOpVectorizationPatterns(
     RewritePatternSet &patterns, PatternBenefit baseBenefit) {
+  // TODO: The following pattern implements "decomposition" and
```
Comment (javedabsar1) [inline mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:2715]: nit. formatting seems wrong. clang-format?

**Review by javedabsar1 — APPROVED** [review]
LGTM. This is wrapping populatePattern and porting test  i.e. mostly NFC.

File: mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp (hunk 4 above)
Comment (banach-space) [reply mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp:2715]: This is meant to match the code below rather than the indented arguments above :)

**Pull request merged**